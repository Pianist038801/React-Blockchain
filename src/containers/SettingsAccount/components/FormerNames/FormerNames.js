import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import Notifications, { notify } from 'react-notify-toast'
import _isNil from 'lodash/isNil'
import classNames from 'classnames'
import Form from 'components/Abstract/Form'
import Datetime from 'react-datetime'

class FormerNames extends Form {
  constructor(props) {
    super(props)
    this.state = {
      errors: [],
      invalidDates: false,
      saving: false,
      first_name: '',
      last_name: '',
      start_date: '',
      end_date: ''
    }
  }

  componentWillReceiveProps(nextProps) {
    const { saving } = this.state
    if (saving && nextProps.saveErrorMsg === null) {
      // new name was added
      this.resetForm()
      this.notifyFn('Name saved correctly', 'success')
    } else if (saving && nextProps.saveErrorMsg !== this.props.saveErrorMsg) {
      // error happened on save
      this.setState({
        saving: false
      })
      this.notifyFn(nextProps.saveErrorMsg, 'error')
    }
  }

  validate() {
    const fields = ['first_name', 'last_name', 'start_date', 'end_date']
    const { start_date, end_date } = this.state
    const errors = []
    let invalidDates = false

    for (let i = 0; i < fields.length; i++) {
      const value = this.state[fields[i]]
      if (_isNil(value) || value === '') {
        errors.push(fields[i])
      }
    }

    if (errors.indexOf('start_date') === -1 &&
      errors.indexOf('end_date') === -1 &&
      new Date(start_date) > new Date(end_date)
    ) {
      invalidDates = true
      errors.push('start_date')
      errors.push('end_date')
    }

    this.setState({
      errors,
      invalidDates
    })

    return { errors, invalidDates }
  }

  onSubmit = (e) => {
    e.preventDefault()
    const { errors, invalidDates } = this.validate()
    const { former_names } = this.props.user
    const { first_name, last_name, start_date, end_date } = this.state
    if (errors.length === 0 && !invalidDates) {
      this.setState({
        saving: true
      })
      former_names.push({ first_name, last_name, start_date, end_date })
      this.props.save({former_names})
    }
  }

  notifyFn(message, type) {
    // this is a wrapper function to make the unit test happy
    notify.show(message, type, 4000)
  }

  resetForm() {
    this.setState({
      errors: [],
      invalidDates: false,
      saving: false,
      first_name: '',
      last_name: '',
      start_date: '',
      end_date: ''
    })
  }

  render() {
    const inputProps = {
      placeholder: 'DD/MM/YYYY'
    }
    const today = moment()
    const validDates = function(current){
      return current.isBefore(today)
    }
    const { pending } = this.props
    const { errors, invalidDates, first_name, last_name, start_date, end_date } = this.state
    const startDate = (start_date !== '') ? moment(start_date) : ''
    const endDate = (end_date !== '') ? moment(end_date) : ''

    return (
      <div>
        <Notifications />
        <h2 className="form-group">Add new former names</h2>
        <form>
          <div className="row">
            <div className="col-md-6">
              <div
                className={classNames('form-group', {
                  'has-error': errors.indexOf('first_name') !== -1
                })}
              >
                <label>First Name</label>
                <input name="first_name" value={first_name} type="text" onChange={this.onChangeInput} />
              </div>
            </div>
            <div className="col-md-6">
              <div
                className={classNames('form-group', {
                  'has-error': errors.indexOf('last_name') !== -1
                })}
              >
                <label>Last Name</label>
                <input name="last_name" value={last_name} type="text" onChange={this.onChangeInput} />
              </div>
            </div>
          </div>

          <div className="row">
            <div className="col-md-6">
              <div
                className={classNames('form-group', {
                  'has-error': errors.indexOf('start_date') !== -1
                })}
              >
                <label>Start Date</label>
                <Datetime
                  name="start_date"
                  value={startDate}
                  dateFormat="DD/MM/YYYY"
                  inputProps={inputProps}
                  timeFormat={false}
                  isValidDate={validDates}
                  onChange={(date) => {
                    if (date && date.format) {
                      this.onChangeInput({
                        target: {
                          name: 'start_date',
                          value: date.format('YYYY-MM-DD')
                        }
                      })
                    }
                  }}
                />
              </div>
            </div>
            <div className="col-md-6">
              <div
                className={classNames('form-group', {
                  'has-error': errors.indexOf('end_date') !== -1
                })}
              >
                <label>End Date</label>
                <Datetime
                  name="end_date"
                  value={endDate}
                  dateFormat="DD/MM/YYYY"
                  inputProps={inputProps}
                  timeFormat={false}
                  isValidDate={validDates}
                  onChange={(date) => {
                    if (date && date.format) {
                      this.onChangeInput({
                        target: {
                          name: 'end_date',
                          value: date.format('YYYY-MM-DD')
                        }
                      })
                    }
                  }}
                />
              </div>
            </div>
          </div>

          {invalidDates &&
            <div className="form-group has-error message">
              End Date can not be earlier than Start Date
            </div>
          }

          <div className="form-group">
            <button
              className="form-btn save"
              disabled={pending}
              onClick={this.onSubmit}
            >
              Add
            </button>
          </div>
        </form>
      </div>
    )
  }
}

FormerNames.propTypes = {
  user: PropTypes.shape({
    former_names: PropTypes.array
  }),
  save: PropTypes.func.isRequired,
  pending: PropTypes.bool.isRequired
}

export default FormerNames
