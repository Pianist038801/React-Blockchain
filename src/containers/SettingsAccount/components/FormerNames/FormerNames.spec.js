import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import FormerNames from './FormerNames'

describe('containers/SettingsAccount/components/FormerNames', () => {
  const props = {
    user: {
      former_names: [
        {first_name: 'jane', last_name: 'doe', start_date: '1990-01-01', end_date: '2016-01-01'}
      ]
    },
    save: jest.fn(),
    pending: false
  }

  const newName = {
    first_name: 'John',
    last_name: 'Doe',
    start_date: '2015-10-30',
    end_date: '2016-09-01'
  }

  describe('when first render', () => {
    it('renders correctly', () => {
      const tree = renderer.create(
        <FormerNames {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when all inputs are invalid', () => {
    it('renders correctly', () => {
      const wrapper = shallow(
        <FormerNames {...props} />
      )
      wrapper.find('button').simulate('click', {preventDefault: jest.fn()})
      const tree = renderer.create(
        wrapper.instance().render()
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  // skip - this fails when running on AWS machine :S
  xdescribe('when end_date is before start_date', () => {
    it('renders correctly', () => {
      const wrapper = shallow(
        <FormerNames {...props} />
      )
      wrapper.setState({
        first_name: 'John',
        last_name: 'Doe',
        end_date: '2016-09-01',
        start_date: '2016-10-30'
      })
      wrapper.find('button').simulate('click', {preventDefault: jest.fn()})
      const tree = renderer.create(
        wrapper.instance().render()
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when all inputs are valid', () => {
    it('called props.save correctly', () => {
      const wrapper = shallow(
        <FormerNames {...props} />
      )
      wrapper.setState(newName)
      wrapper.find('button').simulate('click', {preventDefault: jest.fn()})

      expect(props.save).toHaveBeenCalledWith({
        former_names: [
          props.user.former_names[0],
          newName
        ]
      })
    })
  })

  describe('componentWillReceiveProps', () => {
    describe('when saving is successful', () => {
      let wrapper
      beforeEach(() => {
        wrapper = shallow(
          <FormerNames {...props} />
        )
        jest.spyOn(wrapper.instance(), 'notifyFn').mockImplementation(function() {})
        jest.spyOn(wrapper.instance(), 'resetForm')
        wrapper.setState({saving: true}, function() {
          wrapper.setProps({
            saveErrorMsg: null
          })
        })
      })

      it('calls notifyFn correctly', () => {
        expect(wrapper.instance().notifyFn).toHaveBeenCalledWith(expect.any(String), 'success')
      })

      it('calls resetForm correctly', () => {
        expect(wrapper.instance().resetForm).toHaveBeenCalled()
      })
    })

    describe('when saving is not successful', () => {
      let wrapper
      beforeEach(() => {
        wrapper = shallow(
          <FormerNames {...props} />
        )
        jest.spyOn(wrapper.instance(), 'notifyFn').mockImplementation(function() {})
        wrapper.setState({saving: true}, function() {
          wrapper.setProps({
            saveErrorMsg: 'Error happened'
          })
        })
      })

      it('calls notifyFn correctly', () => {
        expect(wrapper.instance().notifyFn).toHaveBeenCalledWith('Error happened', 'error')
      })

      it('state.saving is set to false', () => {
        expect(wrapper.state('saving')).toEqual(false)
      })
    })
  })
})
