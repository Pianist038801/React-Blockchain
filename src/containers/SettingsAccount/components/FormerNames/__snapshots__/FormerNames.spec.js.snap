// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/SettingsAccount/components/FormerNames when all inputs are invalid renders correctly 1`] = `
<div>
  <div
    id="notification-wrapper"
  />
  <h2
    className="form-group"
  >
    Add new former names
  </h2>
  <form>
    <div
      className="row"
    >
      <div
        className="col-md-6"
      >
        <div
          className="form-group has-error"
        >
          <label>
            First Name
          </label>
          <input
            name="first_name"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="col-md-6"
      >
        <div
          className="form-group has-error"
        >
          <label>
            Last Name
          </label>
          <input
            name="last_name"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-6"
      >
        <div
          className="form-group has-error"
        >
          <label>
            Start Date
          </label>
          <DateTimeMock
            dateFormat="DD/MM/YYYY"
            inputProps={
              Object {
                "placeholder": "DD/MM/YYYY",
              }
            }
            isValidDate={[Function]}
            name="start_date"
            onChange={[Function]}
            timeFormat={false}
            value=""
          />
        </div>
      </div>
      <div
        className="col-md-6"
      >
        <div
          className="form-group has-error"
        >
          <label>
            End Date
          </label>
          <DateTimeMock
            dateFormat="DD/MM/YYYY"
            inputProps={
              Object {
                "placeholder": "DD/MM/YYYY",
              }
            }
            isValidDate={[Function]}
            name="end_date"
            onChange={[Function]}
            timeFormat={false}
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <button
        className="form-btn save"
        disabled={false}
        onClick={[Function]}
      >
        Add
      </button>
    </div>
  </form>
</div>
`;

exports[`containers/SettingsAccount/components/FormerNames when end_date is before start_date renders correctly 1`] = `
<div>
  <div
    id="notification-wrapper"
  />
  <h2
    className="form-group"
  >
    Add new former names
  </h2>
  <form>
    <div
      className="row"
    >
      <div
        className="col-md-6"
      >
        <div
          className="form-group"
        >
          <label>
            First Name
          </label>
          <input
            name="first_name"
            onChange={[Function]}
            type="text"
            value="John"
          />
        </div>
      </div>
      <div
        className="col-md-6"
      >
        <div
          className="form-group"
        >
          <label>
            Last Name
          </label>
          <input
            name="last_name"
            onChange={[Function]}
            type="text"
            value="Doe"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-6"
      >
        <div
          className="form-group has-error"
        >
          <label>
            Start Date
          </label>
          <DateTimeMock
            dateFormat="DD/MM/YYYY"
            inputProps={
              Object {
                "placeholder": "DD/MM/YYYY",
              }
            }
            isValidDate={[Function]}
            name="start_date"
            onChange={[Function]}
            timeFormat={false}
            value={"2016-10-29T23:00:00.000Z"}
          />
        </div>
      </div>
      <div
        className="col-md-6"
      >
        <div
          className="form-group has-error"
        >
          <label>
            End Date
          </label>
          <DateTimeMock
            dateFormat="DD/MM/YYYY"
            inputProps={
              Object {
                "placeholder": "DD/MM/YYYY",
              }
            }
            isValidDate={[Function]}
            name="end_date"
            onChange={[Function]}
            timeFormat={false}
            value={"2016-08-31T23:00:00.000Z"}
          />
        </div>
      </div>
    </div>
    <div
      className="form-group has-error message"
    >
      End Date can not be earlier than Start Date
    </div>
    <div
      className="form-group"
    >
      <button
        className="form-btn save"
        disabled={false}
        onClick={[Function]}
      >
        Add
      </button>
    </div>
  </form>
</div>
`;

exports[`containers/SettingsAccount/components/FormerNames when first render renders correctly 1`] = `
<div>
  <div
    id="notification-wrapper"
  />
  <h2
    className="form-group"
  >
    Add new former names
  </h2>
  <form>
    <div
      className="row"
    >
      <div
        className="col-md-6"
      >
        <div
          className="form-group"
        >
          <label>
            First Name
          </label>
          <input
            name="first_name"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="col-md-6"
      >
        <div
          className="form-group"
        >
          <label>
            Last Name
          </label>
          <input
            name="last_name"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-6"
      >
        <div
          className="form-group"
        >
          <label>
            Start Date
          </label>
          <DateTimeMock
            dateFormat="DD/MM/YYYY"
            inputProps={
              Object {
                "placeholder": "DD/MM/YYYY",
              }
            }
            isValidDate={[Function]}
            name="start_date"
            onChange={[Function]}
            timeFormat={false}
            value=""
          />
        </div>
      </div>
      <div
        className="col-md-6"
      >
        <div
          className="form-group"
        >
          <label>
            End Date
          </label>
          <DateTimeMock
            dateFormat="DD/MM/YYYY"
            inputProps={
              Object {
                "placeholder": "DD/MM/YYYY",
              }
            }
            isValidDate={[Function]}
            name="end_date"
            onChange={[Function]}
            timeFormat={false}
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="form-group"
    >
      <button
        className="form-btn save"
        disabled={false}
        onClick={[Function]}
      >
        Add
      </button>
    </div>
  </form>
</div>
`;
